java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java

```diff
@@ -123,25 +127,69 @@ private static Field getUnionField(Schema schema, String name) {
     int size = schema.getTypes().size();
     long nullCount = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).count();
 
-    // avro schema not allow repeated type, so size == nullCount + 1 indicates nullable type.
-    if (size == nullCount + 1) {
+    // union only has one type, convert to primitive type
+    if (size == 1) {
+      Schema subSchema = schema.getTypes().get(0);
+      return getArrowField(subSchema, name,true);
 
+      // size == 2 and has null type, convert to nullable primitive type
+    } else if (size == 2 && nullCount == 1) {
       Schema nullSchema = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).findFirst().get();
       String nullIndex = String.valueOf(schema.getTypes().indexOf(nullSchema));
+      Schema subSchema = schema.getTypes().stream().filter(s -> s.getType() != Type.NULL).findFirst().get();
+      Preconditions.checkNotNull(subSchema);
+      subSchema.addProp(NULL_INDEX, nullIndex);
```

Comment:
it seems odd to mutate schema here, can you provide more details on why this is necessary?

Reply:
For nullable primitive consumer we need know the null type index in schema, so we could consume a real value or null value from decoder.
Similar for union type, we need know field index for each delegate.
These two data (NULL_INDEX and FIELD_INDEX) is set to avro schema metadata which will transfered in Arrow FieldType, and then we could get it from consumer.

Reply:
can't these be populated directly in the consumers though?  It seems like this information isn't useful beyond the translation process?